### PR TITLE
admin scan endpoint

### PR DESCRIPTION
**Pull request by sdagg9**
finished up code for admin qr scan

### CODE REVIEW

File: src/services/admin/admin-router.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+const adminRouter = Router();
+
+adminRouter.post(
+    "/scan/",
+    RoleChecker([Role.Enum.ADMIN]),
+    async (req, res, next) => {
+        try {
+            const { qrCode } = req.body;
+            if (!qrCode) {
+                console.log("made it to if");
```
Comment (divyack2) [inline src/services/admin/admin-router.ts:20]:
```suggestion
                
```

File: src/services/admin/admin-router.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import { Role } from "../auth/auth-models";
+import dotenv from "dotenv";
+import { validateQrHash } from "./admin-utils";
+
+dotenv.config();
+
+const adminRouter = Router();
+
+adminRouter.post(
+    "/scan/",
```
Comment (divyack2) [inline src/services/admin/admin-router.ts:14]: minor question for @AydanPirani: do we want this to end in a '/' or not? bc some endpoints have the trailing slash and others don't

**Review by divyack2 — APPROVED** [review]
looks good to me! i've just got a quick question for aydan in the comments below but after that you're good to go :))

File: src/services/admin/admin-router.ts (hunk 2 above)
Comment (AydanPirani) [reply src/services/admin/admin-router.ts:14]: ideally not, but it's fine

File: src/services/admin/admin-router.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { Router } from "express";
+import { StatusCodes } from "http-status-codes";
+import { Database } from "../../database";
+import RoleChecker from "../../middleware/role-checker";
+import { Role } from "../auth/auth-models";
+import dotenv from "dotenv";
+import { validateQrHash } from "./admin-utils";
+
+dotenv.config();
```
Comment (AydanPirani) [inline src/services/admin/admin-router.ts:9]:
```suggestion
```

File: src/services/admin/admin-router.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    async (req, res, next) => {
+        try {
+            const { qrCode } = req.body;
+            if (!qrCode) {
+                console.log("made it to if");
+                return res
+                    .status(StatusCodes.BAD_REQUEST)
+                    .json({ error: "QR code is required" });
+            }
+            const { userId, expTime } = validateQrHash(qrCode);
```
Comment (AydanPirani) [inline src/services/admin/admin-router.ts:25]: does the QrHash not contain an eventId as well?

Comment (AydanPirani) [reply src/services/admin/admin-router.ts:25]: or well, our request should contain an eventId and qrCode, can you make a validator in admin service with these 2 fields and verify off that?

File: src/services/admin/admin-utils.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    const hmac = crypto.createHmac("sha256", hashSecret);
+    hashStr = hmac.update(hashStr).digest("hex");
+
+    for (let i = 0; i < hashIterations; i++) {
+        const hash = crypto.createHash("sha256");
+        hashStr = hash.update(hashSecret + "#" + hashStr).digest("hex");
+    }
+
+    return `${hashStr}#${expTime}#${userId}`;
+}
```
Comment (AydanPirani) [inline src/services/admin/admin-utils.ts:18]: just import from attendee/generateQrhash (or export in there if need be)

File: src/services/admin/admin-utils.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    const userId = parts[2];
+    const expTime = parseInt(parts[1]);
+    const generatedHash = generateQrHash(userId, expTime);
+
+    if (generatedHash.split("#")[0] !== parts[0]) {
+        throw new Error("Invalid QR code");
+    }
+
+    return { userId, expTime };
+}
```
Comment (AydanPirani) [inline src/services/admin/admin-utils.ts:31]: also put this in attendee (for now), that way both qr functions are next to each other

**Review by AydanPirani — CHANGES_REQUESTED** [review]

File: src/services/admin/admin-router.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    async (req, res, next) => {
+        try {
+            const { eventId, qrCode } = ScanValidator.parse(req.body);
+            console.log("Event ID:", eventId);
+
+            if (!qrCode) {
+                return res
+                    .status(StatusCodes.BAD_REQUEST)
+                    .json({ error: "QR code is required" });
+            }
```
Comment (AydanPirani) [inline src/services/admin/admin-router.ts:24]:
```suggestion
```

**Review by AydanPirani — APPROVED** [review]
lgtm! one small change but otw all good

File: src/services/checkin/checkin-router.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
+import { Router } from "express";
+import { StatusCodes } from "http-status-codes";
+import { ScanValidator } from "./checkin-schema";
+import { Database } from "../../database";
+import RoleChecker from "../../middleware/role-checker";
+import { Role } from "../auth/auth-models";
+// import dotenv from "dotenv";
```
Comment (AydanPirani) [inline src/services/checkin/checkin-router.ts:7]: remove comment

File: src/app.ts [hunk 9]
```diff
@@ -11,6 +11,7 @@ import bodyParser from "body-parser";
 import errorHandler from "./middleware/error-handler";
 
 import attendeeRouter from "./services/attendee/attendee-router";
+import adminRouter from "./services/checkin/checkin-router";
```
Comment (AydanPirani) [inline src/app.ts:14]: rename to checkinRouter

File: src/app.ts [hunk 10]
```diff
@@ -37,6 +38,7 @@ app.use("/", bodyParser.json());
 
 // API routes
 app.use("/attendee", databaseMiddleware, attendeeRouter);
+app.use("/admin", databaseMiddleware, adminRouter);
```
Comment (AydanPirani) [inline src/app.ts:41]: make this checkin
```suggestion
app.use("/checkin", databaseMiddleware, checkinRouter);
```

Comment (AydanPirani) [reply src/app.ts:41]: (also reorder so alphabetical)

File: src/services/attendee/attendee-utils.ts [hunk 11]
```diff
@@ -16,3 +16,16 @@ export function generateQrHash(userId: string, expTime: number) {
...
+    const userId = parts[2];
+    const expTime = parseInt(parts[1]);
+    const generatedHash = generateQrHash(userId, expTime);
+
+    if (generatedHash.split("#")[0] !== parts[0]) {
+        throw new Error("Invalid QR code");
+    }
+
+    return { userId, expTime };
+}
```
Comment (AydanPirani) [inline src/services/attendee/attendee-utils.ts:31]: move to checkin-utils.ts (both this and generateQrHash)

File: src/services/checkin/checkin-router.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+import { validateQrHash } from "../attendee/attendee-utils";
+import { checkInUserToEvent } from "./checkin-utils";
+
+const adminRouter = Router();
```
Comment (AydanPirani) [inline src/services/checkin/checkin-router.ts:11]: change all occurences of adminRouter to checkinRouter

File: src/services/checkin/checkin-router.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+adminRouter.post(
+    "/checkin/scan/staff",
```
Comment (AydanPirani) [inline src/services/checkin/checkin-router.ts:14]: remove /checkin here
```suggestion
    "/scan/staff",
```

File: src/services/checkin/checkin-router.ts [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+                    .json({ error: "QR code has expired" });
+            }
+
+            const user = await Database.ATTENDEE.findOne({ userId });
+
+            if (!user) {
+                return res
+                    .status(StatusCodes.NOT_FOUND)
+                    .json({ error: "UserNotFound" });
+            }
```
Comment (AydanPirani) [inline src/services/checkin/checkin-router.ts:35]: no need for this, already covered by checkInUserToEvent

**Review by AydanPirani — CHANGES_REQUESTED** [review]
looks good for the most part! a few nits, but otherwise we're good :)

**Review by AydanPirani — APPROVED** [review]
lgtm! ty

**Pull request merged**